### PR TITLE
use dicts for namespaces within text fields

### PR DESCRIPTION
**Pull request by DeNeutoy**

### CODE REVIEW

File: allennlp/data/dataset.py [hunk 1]
```diff
@@ -77,7 +77,8 @@ def get_padding_lengths(self) -> Dict[str, Dict[str, int]]:
 
     def as_arrays(self,
                   padding_lengths: Dict[str, Dict[str, int]] = None,
-                  verbose: bool = True) -> Dict[str, List[numpy.array]]:
+                  verbose: bool = True) -> Dict[str, List[Union[numpy.array,
```
Comment (matt-gardner) [inline allennlp/data/dataset.py]: @joelgrus, is this be another place where generics would work?  I'm not sure if it'd be more work than it's worth.  Basically, each `Field` would return a `DataArray`, which is either a single array, or a dictionary of arrays, and then this type would be `Dict[str, DataArray]`.

Comment (matt-gardner) [inline allennlp/data/dataset.py]: Either way, though, I'm pretty sure the `List` shouldn't be there.  If we stick with the union type, it should just be `Dict[str, Union[numpy.array, Dict[str, numpy.array]]`.

File: allennlp/data/dataset.py [hunk 2]
```diff
@@ -102,12 +103,15 @@ def as_arrays(self,
 
         Returns
         -------
-        data_arrays : ``Dict[str, List[numpy.array]]``
+        data_arrays : ``Dict[str, Union[Dict[str, numpy.array], List[numpy.array]]``
```
Comment (matt-gardner) [inline allennlp/data/dataset.py]: This type should match what's above: `Dict[str, Union[numpy.array, Dict[str, numpy.array]]`.

File: allennlp/data/dataset.py [hunk 3]
```diff
@@ -144,7 +148,17 @@ def as_arrays(self,
...
+                # This is creating a dict of {namespace: batch_array} for each
+                # namespace within this field. This is mostly utilised by TextFields,
+                # which will have one namespace per TokenIndexer used within the Field.
+                namespace_batch_dict = defaultdict(list)
+                for namespace_dict in field_array_list:
+                    for namespace, array in namespace_dict.items():
+                        namespace_batch_dict[namespace].append(array)
+                field_arrays[field_name] = {namespace: numpy.asarray(array_list) for
+                                            namespace, array_list in namespace_batch_dict.items()}
+            elif isinstance(field_array_list[0], (list, tuple)):
```
Comment (matt-gardner) [inline allennlp/data/dataset.py]: I kind of think we should remove this possibility.  Any one `Field` should just return one array, except for `TextField`.

File: allennlp/data/fields/list_field.py [hunk 4]
```diff
@@ -53,12 +53,16 @@ def sequence_length(self) -> int:
         return len(self._field_list)
 
     @overrides
-    def pad(self, padding_lengths: Dict[str, int]) -> List[numpy.array]:
+    def pad(self, padding_lengths: Dict[str, int]) -> Union[Dict[str, numpy.array], List[numpy.array]]:
```
Comment (matt-gardner) [inline allennlp/data/fields/list_field.py]: Can we change this name to `as_array`?

File: allennlp/data/dataset.py (hunk 1 above)
Comment (matt-gardner) [inline allennlp/data/dataset.py]: Also @joelgrus, does mypy now pass on the whole codebase?  If so, can we add it to our travis set up?

File: allennlp/models/simple_tagger.py [hunk 5]
```diff
@@ -54,14 +54,19 @@ def __init__(self,
         self.sequence_loss = torch.nn.CrossEntropyLoss()
 
     # pylint: disable=arguments-differ
-    def forward(self,  # type: ignore
```
Comment (matt-gardner) [inline allennlp/models/simple_tagger.py]: We need this comment.

File: allennlp/models/simple_tagger.py [hunk 6]
```diff
@@ -54,14 +54,19 @@ def __init__(self,
...
+    def forward(self,
+                tokens: Dict[str, torch.LongTensor],
                 tags: torch.LongTensor = None) -> Dict[str, torch.Tensor]:
         """
         Parameters
         ----------
-        tokens : torch.LongTensor, required
-            A torch tensor representing the indices of the
+        tokens : Dict[str, torch.LongTensor], required
+            A dictionary of namespaces which have been indexed to their corresponding tensors.
```
Comment (matt-gardner) [inline allennlp/models/simple_tagger.py]: I would probably start this comment with `The output of TextField.as_array(), which should typically be passed to a TokenEmbedder`, or something similar to that.

File: allennlp/models/simple_tagger.py [hunk 7]
```diff
@@ -127,7 +133,7 @@ def tag(self, text_field: TextField) -> Dict[str, Any]:
         # TODO(Mark): Generalise how the array is transformed into a variable after settling the data API.
         # Add a batch dimension by unsqueezing, because pytorch
         # doesn't support inputs without one.
-        array_input = torch.autograd.Variable(torch.LongTensor(array_input[0])).unsqueeze(0)
+        array_input = {"tokens": torch.autograd.Variable(torch.LongTensor(array_input["tokens"])).unsqueeze(0)}
```
Comment (matt-gardner) [inline allennlp/models/simple_tagger.py]: Hmm, we should have a method somewhere that produces variables from `DataArrays`.  If the array is a dict, we just convert all of the keys.

File: tests/data/dataset_test.py [hunk 8]
```diff
@@ -38,9 +38,9 @@ def test_as_arrays(self):
         dataset.index_instances(self.vocab)
         padding_lengths = dataset.get_padding_lengths()
         arrays = dataset.as_arrays(padding_lengths)
+        text1 = arrays["text1"]["words"]
```
Comment (matt-gardner) [inline tests/data/dataset_test.py]: Why did you change the namespace from "tokens" to "words" in the test?  I think it'd be better to have it use the defaults - reading through this test might give you the wrong impression about how you should access the arrays.

Comment (DeNeutoy) [inline tests/data/dataset_test.py]: I kind of disagree - I think it's helpful to have tests which show how some features are connected together and this is well within the realm of what we should be able to do with a `Vocab`, `Dataset` and `TokenIndexer`. Also I don't exactly understand what you mean about how this gives the wrong impression of how to access the arrays?

File: allennlp/data/dataset.py (hunk 3 above)
Comment (DeNeutoy) [inline allennlp/data/dataset.py]: Yeah I left this as it's still possible that `ListField` would yield lists, but really I don't think there's any reason to keep this. I'll change ListField.

File: tests/data/dataset_test.py (hunk 8 above)
Comment (matt-gardner) [inline tests/data/dataset_test.py]: My issue was that I was reading through this test and wondered where "words" came from.  It's not obvious, and it's not the default.  I only know it's not the default because I made the default - someone who is just looking at the test might think that "words" is what you should normally use, because there's nothing obvious in this test that's changing something to "words".

File: allennlp/models/simple_tagger.py (hunk 7 above)
Comment (DeNeutoy) [inline allennlp/models/simple_tagger.py]: Yeah that's what this TODO above is about - I'm leaving it for now because it should also move tensors to GPUs and so needs to be a little more complicated than otherwise it might be.

File: allennlp/data/dataset.py (hunk 1 above)
Comment (joelgrus) [inline allennlp/data/dataset.py]: (1) yes, I think generics would work here in the same way, it shouldn't be too much work
(2) yes, as of my last commit, mypy works on the whole codebase, I'll add it to travis (I might need to add it to `requirements.txt` too

File: tests/data/fields/text_field_test.py [hunk 9]
```diff
@@ -88,19 +88,19 @@ def test_index_converts_field_correctly(self):
         t_index = vocab.add_token_to_namespace("t", namespace='characters')
         c_index = vocab.add_token_to_namespace("c", namespace='characters')
 
-        field = TextField(["A", "sentence"], [token_indexers["single id"](token_namespace="words")])
+        field = TextField(["A", "sentence"], [token_indexers["single id"](namespace="words")])
         field.index(vocab)
         # pylint: disable=protected-access
         assert field._indexed_tokens == [[capital_a_index, sentence_index]]
```
Comment (matt-gardner) [inline tests/data/fields/text_field_test.py]: I was looking at this, and wondered why this still passed.  Then I realized that even though the return value of `as_array` is a `Dict`, `token_indexers` is still a list, and we have the issue of how to map from this list to keys in the `Dict`.  Why don't we just make the `token_indexers` a dict, and the user assigns their own keys?  I think that would clear up the issue with using "words" vs. "tokens", too.  What do you think?

Comment (DeNeutoy) [inline tests/data/fields/text_field_test.py]: No - I unified the `namespace` attribute of the `TokenIndexer` so we can use it here: https://github.com/allenai/allennlp/pull/25/files#diff-ad92587fe15aa343c6d63c5f1c06cf5fR88. Your way is another solution but this is better because it more directly links using a `TokenIndexer` with a same vocabulary namespace. Actually we should raise an error somewhere if you try to index a single `TextField` with a `TokenIndexer` using the same vocab namespace. I'll add that.

Comment (matt-gardner) [inline tests/data/fields/text_field_test.py]: I'm wondering if directly linking those two is desirable, and I'm thinking it's not (I thought you also had some concerns about this when we first talked about it).  The vocab namespace is something that you shouldn't have to think about at all except for rare edge cases, but you have to if these are linked.  If we instead just have arbitrary keys, it's really easy to make sure they match up in the configuration file:

```json
{
  "token_indexers": {
    "words": "single_token",
    "characters": "token_characters",
  },
  ...
  "token embedder": {
    "words": "single_token_embedding",
    "characters": {
      "type": "cnn"
    }
  }
}
```

(some details missing, but I think you get the idea)

If you don't use a dictionary to do this, the keys that a token embedder should look for look like magic.

Comment (matt-gardner) [inline tests/data/fields/text_field_test.py]: And, if I want to be really crazy, I can do that too:

```json
{
  "token indexers": {
    "words": "single_token",
    "tf-idf": "single_token"
  },
  "token embedder": {
    "words": "glove vectors",
    "tf-idf": "single_token_embedding with dim 1"
  }
}
```

Comment (DeNeutoy) [inline tests/data/fields/text_field_test.py]: You're right, as usual.

File: allennlp/data/fields/sequence_field.py [hunk 10]
```diff
@@ -1,7 +1,8 @@
-from allennlp.data.fields.field import Field
+from typing import Generic
+from allennlp.data.fields.field import Field, DataArray
 
 
-class SequenceField(Field):
+class SequenceField(Field[DataArray], Generic[DataArray]):
```
Comment (matt-gardner) [inline allennlp/data/fields/sequence_field.py]: Why does this need to inherit from both?

Comment (joelgrus) [inline allennlp/data/fields/sequence_field.py]: this seems to be the way to have two levels of generic abstract classes. see e.g.

https://github.com/python/mypy/issues/2291

Comment (joelgrus) [inline allennlp/data/fields/sequence_field.py]: actually, I take that back, that issue is closed, it should work with just the first one

https://github.com/python/mypy/pull/2811

@DeNeutoy sorry for directing you wrong

**Comment by DeNeutoy** [comment]
@matt-gardner Sorry this turned out to be so massive - when I moved the `TokenIndexers` over to be dicts in the `TextFields`, I realised it made sense for the `self._indexed_tokens` to also be a dict rather than a list (it actually had to be as otherwise we can't align the indexers to the list), which made this a bit bigger.

File: allennlp/data/dataset.py [hunk 11]
```diff
@@ -152,11 +151,11 @@ def as_arrays(self,  # ignore type
                 # This is creating a dict of {namespace: batch_array} for each
                 # namespace within this field. This is mostly utilised by TextFields,
                 # which will have one namespace per TokenIndexer used within the Field.
-                namespace_batch_dict = defaultdict(list)
+                namespace_batch_dict = defaultdict(list)  # type: Dict[str, List[numpy.array]]
                 for namespace_dict in field_array_list:
                     for namespace, array in namespace_dict.items():
                         namespace_batch_dict[namespace].append(array)
-                field_arrays[field_name] = {namespace: numpy.asarray(array_list) for
+                field_arrays[field_name] = {namespace: numpy.asarray(array_list) for  # type: ignore
```
Comment (DeNeutoy) [inline allennlp/data/dataset.py]: @joelgrus - I ignored the type here, because we are modifying this dictionary in place and it starts off with a different type for its values. Maybe I could change the type above to be a union of the possible types, is that better?

Comment (matt-gardner) [inline allennlp/data/dataset.py]: Or just create a new variable for the final return value?  That would remove the need for the `{**field_arrays}` magic, too.

Comment (DeNeutoy) [inline allennlp/data/dataset.py]: I assumed that this was for memory efficiency, so we don't create 2 copies of a large dataset though? Python can't garbage collect the first dict until we are outside this loop, right?

Comment (matt-gardner) [inline allennlp/data/dataset.py]: If we're worried about early garbage collection, we can just do:

```python
final_arrays = {}
for field_name, field_array_list in field_arrays.items():
    # logic for dict here
    final_arrays[field_name] = numpy.asarray(field_array_list)
    del field_arrays[field_name]
```

Comment (DeNeutoy) [inline allennlp/data/dataset.py]: Ok, sure.

**Comment by DeNeutoy** [comment]
Oh ok, we actually can't change the size mid iteration, I forgot that.

File: allennlp/data/dataset_readers/sequence_tagging.py [hunk 12]
```diff
@@ -53,11 +53,15 @@ def from_params(cls, params: Params):
...
         """
         filename = params.pop('filename')
-        token_indexers = [TokenIndexer.from_params(p)
-                          for p in params.pop('token_indexers', [Params({})])]
+        token_indexers = {}
+        token_indexer_params = params.pop('token_indexers', Params({}))
+        for name, indexer_params in token_indexer_params.items():
+            token_indexers[name] = TokenIndexer.from_params(indexer_params)
+        if token_indexers == {}:
+            token_indexers = None
```
Comment (matt-gardner) [inline allennlp/data/dataset_readers/sequence_tagging.py]: That same comment would probably be useful here, too.

File: allennlp/data/dataset_readers/snli.py [hunk 13]
```diff
@@ -66,8 +64,12 @@ def from_params(cls, params: Params):
...
         filename = params.pop('filename')
         tokenizer = Tokenizer.from_params(params.pop('tokenizer', {}))
-        token_indexers = [TokenIndexer.from_params(p)
-                          for p in params.pop('token_indexers', [Params({})])]
+        token_indexers = {}
+        token_indexer_params = params.pop('token_indexers', Params({}))
+        for name, indexer_params in token_indexer_params.items():
+            token_indexers[name] = TokenIndexer.from_params(indexer_params)
+        if token_indexers == {}:
+            token_indexers = None
```
Comment (matt-gardner) [inline allennlp/data/dataset_readers/snli.py]: And here.

File: allennlp/data/dataset_readers/squad.py [hunk 14]
```diff
@@ -233,8 +231,12 @@ def from_params(cls, params: Params):
...
         negative_sentence_selection = params.pop('negative_sentence_selection', 'paragraph')
         tokenizer = Tokenizer.from_params(params.pop('tokenizer', {}))
-        token_indexers = [TokenIndexer.from_params(p)
-                          for p in params.pop('token_indexers', [Params({})])]
+        token_indexers = {}
+        token_indexer_params = params.pop('token_indexers', Params({}))
+        for name, indexer_params in token_indexer_params.items():
+            token_indexers[name] = TokenIndexer.from_params(indexer_params)
+        if token_indexers == {}:
+            token_indexers = None
```
Comment (matt-gardner) [inline allennlp/data/dataset_readers/squad.py]: And here...  Seems a bit redundant to have it everywhere, but the problem shows up when you read the first file, and we don't know which one you're going to read first.

File: allennlp/data/fields/list_field.py [hunk 15]
```diff
@@ -24,7 +25,7 @@ class ListField(SequenceField):
         A list of ``Field`` objects to be concatenated into a single input tensor.  All of the
         contained ``Field`` objects must be of the same type.
     """
-    def __init__(self, field_list: List[Field]) -> None:
+    def __init__(self, field_list: List[Field]) -> None:  # pylint: disable=super-init-not-called
```
Comment (matt-gardner) [inline allennlp/data/fields/list_field.py]: Is this pylint comment actually necessary?  There's no init in `SequenceField`.  Seems odd...

Comment (matt-gardner) [inline allennlp/data/fields/list_field.py]: Oh, looks like this should be `ListField(SequenceField[DataArray])`.  Maybe that would fix this error?

File: allennlp/data/fields/text_field.py [hunk 16]
```diff
@@ -14,50 +14,51 @@
...
     to tokenize raw strings using a :class:`..tokenizers.Tokenizer`.
 
     Because string tokens can be represented as indexed arrays in a number of ways, we also take a
-    list of :class:`TokenIndexer` objects that will be used to convert the tokens into indices.
+    dictionary of :class:`TokenIndexer` objects that will be used to convert the tokens into indices.
     Each ``TokenIndexer`` could represent each token as a single ID, or a list of character IDs, or
     something else.
 
-    This field will get converted into a list of arrays, one for each ``TokenIndexer``.  A
+    This field will get converted into apy dictionary of arrays, one for each ``TokenIndexer``.  A
```
Comment (matt-gardner) [inline allennlp/data/fields/text_field.py]: Typo here.

File: allennlp/data/fields/text_field.py [hunk 17]
```diff
@@ -14,50 +14,51 @@
...
             raise ConfigurationError("You must call .index(vocabulary) on a "
                                      "field before determining padding lengths.")
-        for indexer, array in zip(self._token_indexers, self._indexed_tokens):
+        for indexer_name, indexer in self._token_indexers.items():
             indexer_lengths = {}
 
             # This is a list of dicts, one for each token in the field.
-            token_lengths = [indexer.get_padding_lengths(token) for token in array]
+            token_lengths = [indexer.get_padding_lengths(token) for token in self._indexed_tokens[indexer_name]]
             # TODO(Mark): This breaks if the token list is empty, but we need to be able to have empty fields.
```
Comment (matt-gardner) [inline allennlp/data/fields/text_field.py]: Not sure what this is talking about.  You mean the `self.empty_field()` method?  Because of where `empty_field()` is called, you're guaranteed to never have this method called on the result of `get_padding_lengths`.

File: allennlp/data/fields/text_field.py [hunk 18]
```diff
@@ -66,33 +67,37 @@ def get_padding_lengths(self) -> Dict[str, int]:
...
-    def pad(self, padding_lengths: Dict[str, int]) -> List[numpy.array]:
-        arrays = []
+    # @overrides
+    def as_array(self, padding_lengths: Dict[str, int]) -> Dict[str, numpy.array]:
+        arrays = {}
         desired_num_tokens = padding_lengths['num_tokens']
-        for indexer, array in zip(self._token_indexers, self._indexed_tokens):
-            padded_array = indexer.pad_token_sequence(array, desired_num_tokens, padding_lengths)
-            arrays.append(numpy.asarray(padded_array))
+        for (indexer_name, indexer) in self._token_indexers.items():
```
Comment (matt-gardner) [inline allennlp/data/fields/text_field.py]: Parens around here aren't necessary - do you find them more readable?  I think we haven't used them, generally.  (I don't have a strong opinion either way, just wondering if we should pick one way and be consistent.)

File: allennlp/data/fields/text_field.py [hunk 19]
```diff
@@ -66,33 +67,37 @@ def get_padding_lengths(self) -> Dict[str, int]:
...
+            padded_array = indexer.pad_token_sequence(self._indexed_tokens[indexer_name],
+                                                      desired_num_tokens, padding_lengths)
+            # Use the namespace of the indexer as a key to recognise what
```
Comment (matt-gardner) [inline allennlp/data/fields/text_field.py]: I'd avoid the word "namespace" here - maybe just say "We use the same keys in the array dictionary as were used for the token indexers given in the constructor", or something similar.

File: allennlp/data/fields/text_field.py [hunk 20]
```diff
@@ -66,33 +67,37 @@ def get_padding_lengths(self) -> Dict[str, int]:
...
+            # the array corresponds to in a model.
+            arrays[indexer_name] = padded_array
         return arrays
 
-    @overrides
+    # @overrides
     def empty_field(self):
         # pylint: disable=protected-access
         text_field = TextField([], self._token_indexers)
         # This needs to be a list of empty lists for each token_indexer,
```
Comment (matt-gardner) [inline allennlp/data/fields/text_field.py]: "needs to be a dict of empty lists"

File: tests/data/dataset_test.py [hunk 21]
```diff
@@ -14,16 +14,16 @@
 class TestDataset(AllenNlpTestCase):
     def setUp(self):
         self.vocab = Vocabulary()
-        self.vocab.add_token_to_namespace("this", "words")
-        self.vocab.add_token_to_namespace("is", "words")
-        self.vocab.add_token_to_namespace("a", "words")
-        self.vocab.add_token_to_namespace("sentence", "words")
-        self.vocab.add_token_to_namespace(".", "words")
+        self.vocab.add_token_to_namespace("this")
```
Comment (matt-gardner) [inline tests/data/dataset_test.py]: Unrelated to this PR - I see this line and I think "add_token_to_namespace" is too verbose.  Would it be problematic to just change this to `add_token`, using the `namespace=''` argument when necessary?

File: tests/data/dataset_test.py [hunk 22]
```diff
@@ -38,22 +38,22 @@ def test_as_arrays(self):
...
-        text2 = arrays["text2"][0]
         numpy.testing.assert_array_almost_equal(text1, numpy.array([[2, 3, 4, 5, 6],
                                                                     [1, 3, 4, 5, 6]]))
         numpy.testing.assert_array_almost_equal(text2, numpy.array([[2, 3, 4, 1, 5, 6],
                                                                     [2, 3, 1, 0, 0, 0]]))
 
     def get_dataset(self):
         field1 = TextField(["this", "is", "a", "sentence", "."],
-                           [token_indexers["single id"]("words")])
+                           {"tokens": token_indexers["single id"]()})
```
Comment (matt-gardner) [inline tests/data/dataset_test.py]: You could just define this once, at the top of the method, and use the variable in these four lines.

File: tests/data/fields/list_field_test.py [hunk 23]
```diff
@@ -23,9 +23,12 @@ def setUp(self):
         super(TestListField, self).setUp()
 
     def test_get_padding_lengths(self):
-        field1 = TextField(["this", "is", "a", "sentence"], [token_indexers["single id"]("words")])
-        field2 = TextField(["this", "is", "a", "different", "sentence"], [token_indexers["single id"]("words")])
-        field3 = TextField(["this", "is", "another", "sentence"], [token_indexers["single id"]("words")])
+        field1 = TextField(["this", "is", "a", "sentence"],
+                           {"words": token_indexers["single id"]("words")})
```
Comment (matt-gardner) [inline tests/data/fields/list_field_test.py]: I'd also recommend defining this once at the top of the method.

File: tests/data/fields/list_field_test.py [hunk 24]
```diff
@@ -34,52 +37,59 @@ def test_get_padding_lengths(self):
         assert lengths == {"num_fields": 3, "num_tokens": 5}
 
     def test_all_fields_padded_to_max_length(self):
-        field1 = TextField(["this", "is", "a", "sentence"], [token_indexers["single id"]("words")])
-        field2 = TextField(["this", "is", "a", "different", "sentence"], [token_indexers["single id"]("words")])
-        field3 = TextField(["this", "is", "another", "sentence"], [token_indexers["single id"]("words")])
+        field1 = TextField(["this", "is", "a", "sentence"],
+                           {"words": token_indexers["single id"]("words")})
```
Comment (matt-gardner) [inline tests/data/fields/list_field_test.py]: And here.

File: tests/data/fields/list_field_test.py [hunk 25]
```diff
@@ -34,52 +37,59 @@ def test_get_padding_lengths(self):
...
+        numpy.testing.assert_array_almost_equal(array_dict["words"][1], numpy.array([2, 3, 4, 1, 5]))
+        numpy.testing.assert_array_almost_equal(array_dict["words"][2], numpy.array([2, 3, 1, 5, 0]))
 
     def test_fields_can_pad_to_greater_than_max_length(self):
 
-        field1 = TextField(["this", "is", "a", "sentence"], [token_indexers["single id"]("words")])
-        field2 = TextField(["this", "is", "a", "different", "sentence"], [token_indexers["single id"]("words")])
-        field3 = TextField(["this", "is", "another", "sentence"], [token_indexers["single id"]("words")])
+        field1 = TextField(["this", "is", "a", "sentence"],
+                           {"words": token_indexers["single id"]("words")})
```
Comment (matt-gardner) [inline tests/data/fields/list_field_test.py]: Here.

File: tests/data/fields/list_field_test.py [hunk 26]
```diff
@@ -34,52 +37,59 @@ def test_get_padding_lengths(self):
...
+        numpy.testing.assert_array_almost_equal(array_dict["words"][1], numpy.array([2, 3, 4, 1, 5, 0, 0]))
+        numpy.testing.assert_array_almost_equal(array_dict["words"][2], numpy.array([2, 3, 1, 5, 0, 0, 0]))
+        numpy.testing.assert_array_almost_equal(array_dict["words"][3], numpy.array([0, 0, 0, 0, 0, 0, 0]))
+        numpy.testing.assert_array_almost_equal(array_dict["words"][4], numpy.array([0, 0, 0, 0, 0, 0, 0]))
 
     def test_pad_can_handle_multiple_token_indexers(self):
         field1 = TextField(["this", "is", "a", "sentence"],
-                           [token_indexers["single id"]("words"),
-                            token_indexers["characters"]("characters")])
+                           {"words": token_indexers["single id"]("words"),
```
Comment (matt-gardner) [inline tests/data/fields/list_field_test.py]: Here.

File: allennlp/data/dataset.py [hunk 27]
```diff
@@ -77,7 +77,7 @@ def get_padding_lengths(self) -> Dict[str, Dict[str, int]]:
 
     def as_arrays(self,
                   padding_lengths: Dict[str, Dict[str, int]] = None,
-                  verbose: bool = True) -> Dict[str, List[numpy.array]]:
+                  verbose: bool = True) -> Dict[str, Union[numpy.array, Dict[str, numpy.array]]]:
```
Comment (DeNeutoy) [inline allennlp/data/dataset.py]: It should, but it broke the type checking 😢 . I discussed this with @joelgrus  and we decided it wasn't worth the time figuring out why this was happening - it was something to do with using the `TypeVar` annotation from `typing`, because defining a new type just equal to this complex type didn't break mypy.

Comment (matt-gardner) [inline allennlp/data/dataset.py]: That's too bad.  Maybe leave a comment about the type, then?

File: allennlp/data/fields/list_field.py (hunk 15 above)
Comment (DeNeutoy) [inline allennlp/data/fields/list_field.py]: You're right that fixed it .... The way these types interact with pylint seems a bit magical to me.

File: tests/data/dataset_test.py (hunk 21 above)
Comment (DeNeutoy) [inline tests/data/dataset_test.py]: Possibly, i'll ask Nick F and some Aristo people today.

File: allennlp/data/dataset.py (hunk 27 above)
Comment (DeNeutoy) [inline allennlp/data/dataset.py]: I'm just going to change it and ignore the type, actually.

File: allennlp/data/fields/text_field.py (hunk 17 above)
Comment (DeNeutoy) [inline allennlp/data/fields/text_field.py]: No, the problem occurs when we pad a `ListField` to a maximum length with empty `TextFields`, because we then call get padding lengths with a `TextField` containing no tokens, but we expect below (`for key in token_lengths[0].keys()`) that there is a minimum of 1 element in `token_fields`. I left this previously but I think I can fix it now.

Comment (DeNeutoy) [inline allennlp/data/fields/text_field.py]: Actually, no, this is non-trivial, i'm going to leave it for another PR.
